executor/worker/task_committer.go

```diff
@@ -0,0 +1,203 @@
+package worker
+
+import (
+	"sync"
+	"time"
+
+	"github.com/pingcap/tiflow/dm/pkg/log"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+
+	"github.com/hanfei1991/microcosm/executor/worker/internal"
+	"github.com/hanfei1991/microcosm/pkg/clock"
+)
+
+const runTTLCheckerInterval = 1 * time.Second
+
+type requestID = string
+
+type requestEntry struct {
+	RequestID requestID
+	ExpireAt  time.Time
+
+	// task is a RunnableContainer, which contains
+	// the task's submit time.
+	task *internal.RunnableContainer
+}
+
+func (r *requestEntry) TaskID() internal.RunnableID {
+	return r.task.ID()
+}
+
+// WrappedTaskAdder is an interface used to abstract a TaskRunner.
+type WrappedTaskAdder interface {
+	addWrappedTask(task *internal.RunnableContainer) error
+}
+
+// TaskCommitter is used to implement two-phase task dispatching.
+type TaskCommitter struct {
+	runner WrappedTaskAdder
+
+	mu               sync.Mutex
+	pendingRequests  map[requestID]*requestEntry
+	requestsByTaskID map[RunnableID]*requestEntry
+
+	clock    clock.Clock
+	wg       sync.WaitGroup
+	cancelCh chan struct{}
+
+	// for unit tests only
+	requestCleanUpCount atomic.Int64
+
+	requestTTL time.Duration
+}
+
+// NewTaskCommitter returns a TaskCommitter.
+func NewTaskCommitter(runner WrappedTaskAdder, requestTTL time.Duration) *TaskCommitter {
+	return newTaskCommitterWithClock(runner, requestTTL, clock.New())
+}
+
+func newTaskCommitterWithClock(
+	runner WrappedTaskAdder,
+	requestTTL time.Duration,
+	clock clock.Clock,
+) *TaskCommitter {
+	committer := &TaskCommitter{
+		runner: runner,
+
+		pendingRequests:  make(map[requestID]*requestEntry),
+		requestsByTaskID: make(map[RunnableID]*requestEntry),
+
+		clock:      clock,
+		cancelCh:   make(chan struct{}),
+		requestTTL: requestTTL,
+	}
+
+	committer.wg.Add(1)
+	go func() {
+		defer committer.wg.Done()
+		committer.runTTLChecker()
+	}()
+
+	return committer
+}
+
+// PreDispatchTask is the "prepare" stage of submitting a task.
+func (c *TaskCommitter) PreDispatchTask(rID requestID, task Runnable) (ok bool) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if _, exists := c.pendingRequests[rID]; exists {
+		// Duplicate requests are not allowed.
+		// As we use UUIDs as request IDs, there should not be duplications.
+		// The calling convention is that you should NOT retry the pre-dispatch API call.
+		return false
+	}
+
+	// We need to overwrite stale request for the same workerID.
+	if request, exists := c.requestsByTaskID[task.ID()]; exists {
+		log.L().Info("There exists a previous request with the same worker ID, overwriting it.",
+			zap.Any("request", request))
+
+		c.removeRequestByID(request.RequestID)
+	}
+
+	// We use the current time as the submit time of the task.
+	taskWithSubmitTime := internal.WrapRunnable(task, c.clock.Now())
+
+	request := &requestEntry{
+		RequestID: rID,
+		ExpireAt:  c.clock.Now().Add(c.requestTTL),
```

Comment:
Do we have different expectation of the requestTTL and taskWithSubmitTime?
In detail, should requestTTL be shorter than the worker timeout?

Reply:
It does not matter. The client should not know about `requestTTL`, and if it wants to recreate a worker with the same name, the old request will be rewritten.

Reply:
We trust that the worker can cancel itself if it has timed out. We do not need another mechanism to prevent a worker from being run.